cirq/ionq/serializer.py

```diff
@@ -50,46 +65,84 @@ def __init__(self, atol: float = 1e-8):
             common_gates.CNotPowGate: self._serialize_cnot_pow_gate,
             common_gates.HPowGate: self._serialize_h_pow_gate,
             common_gates.SwapPowGate: self._serialize_swap_gate,
+            common_gates.MeasurementGate: self._serialize_measurement_gate,
         }
 
-    def serialize(self, circuit: 'cirq.Circuit') -> dict:
+    def serialize(self, circuit: 'cirq.Circuit') -> SerializedProgram:
         """Serialize the given circuit.
 
         Raises:
-            ValueError: if the circuit has gates that are not supported or
-                is otherwise invalid.
+            ValueError: if the circuit has gates that are not supported or is otherwise invalid.
         """
+        self._validate_circuit(circuit)
+        num_qubits = self._validate_qubits(circuit.all_qubits())
+
+        serialized_ops = self._serialize_circuit(circuit)
+
+        # IonQ API does not support measurements, so we pass the measurement keys through
+        # the metadata field.  Here we split these out of the serialized ops.
+        body = {
+            'qubits': num_qubits,
+            'circuit': [op for op in serialized_ops if op['gate'] != 'meas'],
+        }
+        metadata = {op['key']: op['targets'] for op in serialized_ops if op['gate'] == 'meas'}
+        self._validate_metadata(metadata)
+        return SerializedProgram(body=body, metadata=metadata)
+
+    def _validate_circuit(self, circuit: 'cirq.Circuit'):
         if len(circuit) == 0:
             raise ValueError('Cannot serialize empty circuit.')
-        all_qubits = circuit.all_qubits()
+        if not circuit.are_all_measurements_terminal():
+            raise ValueError('All measurements in circuit must be at end of circuit.')
+
+    def _validate_qubits(self, all_qubits: Collection['cirq.Qid']) -> int:
+        """Returns the number of qubits while validating qubit types and values."""
         if any(not isinstance(q, line_qubit.LineQubit) for q in all_qubits):
             raise ValueError(
                 f'All qubits must be cirq.LineQubits but were {set(type(q) for q in all_qubits)}'
             )
+        if any(cast(line_qubit.LineQubit, q).x < 0 for q in all_qubits):
+            raise ValueError(
+                'IonQ API must use LineQubits from 0 to number of qubits - 1. Instead found line '
+                f'qubits with indices {all_qubits}.'
+            )
         num_qubits = cast(line_qubit.LineQubit, max(all_qubits)).x + 1
-        return {'qubits': num_qubits, 'circuit': self._serialize_circuit(circuit, num_qubits)}
+        return num_qubits
 
-    def _serialize_circuit(self, circuit: 'cirq.Circuit', num_qubits: int) -> list:
+    def _validate_metadata(self, metadata: dict):
+        if len(metadata) > 9:
```

Comment:
I guess this also depends on other fields added to the metadata dict. Another reason to put measurement keys under specific metadata dict key, so we only check size on that one.

Reply:
Serialized across potentially multiple keys, so made that check here.